qa/tasks/ceph_test_case.py

```diff
@@ -118,16 +121,25 @@ def __exit__(self, exc_type, exc_val, exc_tb):
                         # then some grace
                         time.sleep(5 + timeout)
 
-                self.watcher_process.stdin.close()
+                traceback.print_exc()
+                log.info('=============will close self.watcher_process.stdin')
+                rv = self.watcher_process.stdin.close()
```

Comment:
So the reason your `shell=False` works is because closing `stdin` will then send `SIGTERM` to the `ceph -w` process rather than the shell. Killing the shell does not kill its children. `ceph -w` does not read from `stdin` so closing this channel does nothing to cause the `ceph -w` process to exit.

What would probably help here is to also close `self.watcher_process.stdout`. Then `ceph -w` would receive `SIGPIPE` when it next tries to write to `stdout`. You need to first read whatever data is on `stdout` before closing it however. IOW, move the `self.match()` check here, then close, then wait.

Process management is fun, huh? :)

Reply:
Ah, that explanation was **really** helpful. Thanks!  :D

> IOW, move the self.match() check here, then close, then wait.

Which one? The one at the top or the one at the bottom? I think the one at the top is not related in our context and one at the bottom is perfect where it is since self.match() just might find the pattern when stdout is written just before closing it.

Reply:
I verified it by moving the self.match at the bottom to just before the stdout and stderr is closed -

```
020-12-08 23:12:36,637.637 INFO:__main__:======================================================================
2020-12-08 23:12:36,637.637 INFO:__main__:FAIL: test_selftest_cluster_log (tasks.mgr.test_module_selftest.TestModuleSelftest)
2020-12-08 23:12:36,638.638 INFO:__main__:Use the selftest module to test the cluster/audit log interface.
2020-12-08 23:12:36,638.638 INFO:__main__:
2020-12-08 23:12:36,638.638 INFO:__main__:----------------------------------------------------------------------
2020-12-08 23:12:36,638.638 INFO:__main__:Traceback (most recent call last):
2020-12-08 23:12:36,638.638 INFO:__main__:  File "/home/rishabh/repos/ceph/minor/qa/tasks/mgr/test_module_selftest.py", line 242, in test_selftest_cluster_log
2020-12-08 23:12:36,638.638 INFO:__main__:    log.info('============rv = ' + str(rv))
2020-12-08 23:12:36,638.638 INFO:__main__:  File "/home/rishabh/repos/ceph/minor/qa/tasks/ceph_test_case.py", line 127, in __exit__
2020-12-08 23:12:36,638.638 INFO:__main__:    raise AssertionError("Expected log message not found: '{0}'".format(expected_pattern))
2020-12-08 23:12:36,638.638 INFO:__main__:AssertionError: Expected log message not found: '[INF] foo bar info'
2020-12-08 23:12:36,638.638 INFO:__main__:
```


```
$ git diff
diff --git a/qa/tasks/ceph_test_case.py b/qa/tasks/ceph_test_case.py
index 7d40532854..9e64aa88f7 100644
--- a/qa/tasks/ceph_test_case.py
+++ b/qa/tasks/ceph_test_case.py
@@ -121,6 +121,10 @@ class CephTestCase(unittest.TestCase):
                         # then some grace
                         time.sleep(5 + timeout)
 
+                if not self.match():
+                    log.error("Log output: \n{0}\n".format(self.watcher_process.stdout.getvalue()))
+                    raise AssertionError("Expected log message not found: '{0}'".format(expected_pattern))
+
                 traceback.print_exc()
                 log.info('=============will close self.watcher_process.stdout and stderr')
                 self.watcher_process.close_stdout_and_stderr()
@@ -137,10 +141,6 @@ class CephTestCase(unittest.TestCase):
                     log.info('===========e.args' + str(e.args))
                     pass
 
-                if not self.match():
-                    log.error("Log output: \n{0}\n".format(self.watcher_process.stdout.getvalue()))
-                    raise AssertionError("Expected log message not found: '{0}'".format(expected_pattern))
-
                 log.info('==============finished running assert_cluster_log.ContextManager.__exit__()')
 
         return ContextManager()
```

Reply:
Updating the branch right now without moving self.match().

Reply:
> So the reason your shell=False works is because closing stdin will then send SIGTERM to the ceph -w process rather than the shell. Killing the shell does not kill its children. ceph -w does not read from stdin so closing this channel does nothing to cause the ceph -w process to exit.

On second thought, why would the crash happen only on CI? Locally, the test always ran successfully.

Reply:
@batrick, forgot to tag you here - https://github.com/ceph/ceph/pull/38471#discussion_r538997245

Reply:
Perhaps because logging is much faster locally.

Reply:
That's true, I've seen it myself but how does that affect the crashing? How is it related to launching and killing of shell and it's child process?

Reply:
There is a "match" line before closing stdin. That probably gets hit every time in local testing.